Build error on GHC 8.0 and 8.2

Here is a build log when attempting to build `ordered-containers-0.2` with GHC 8.2.2:

```
Building library for ordered-containers-0.2..
[1 of 3] Compiling Data.Map.Util    ( Data/Map/Util.hs, dist/build/Data/Map/Util.o )
[2 of 3] Compiling Data.Map.Ordered ( Data/Map/Ordered.hs, dist/build/Data/Map/Ordered.o )

Data/Map/Ordered.hs:80:53: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Data.Monoid.<>’,
                             imported from ‘Data.Monoid’ at Data/Map/Ordered.hs:44:1-18
                          or ‘Data.Semigroup.<>’,
                             imported from ‘Data.Semigroup’ at Data/Map/Ordered.hs:46:1-21
   |
80 |         Bias o <> Bias o' = Bias (unionWithL (const (<>)) o o')
   |                                                     ^^^^

Data/Map/Ordered.hs:82:53: error:
    Ambiguous occurrence ‘<>’
    It could refer to either ‘Data.Monoid.<>’,
                             imported from ‘Data.Monoid’ at Data/Map/Ordered.hs:44:1-18
                          or ‘Data.Semigroup.<>’,
                             imported from ‘Data.Semigroup’ at Data/Map/Ordered.hs:46:1-21
   |
82 |         Bias o <> Bias o' = Bias (unionWithR (const (<>)) o o')
   |                                                     ^^^^
```

--- a/Data/Map/Ordered.hs
+++ b/Data/Map/Ordered.hs
@@ -41,9 +41,13 @@ import Data.Foldable (Foldable, foldl', foldMap)
 import Data.Function (on)
 import Data.Map (Map)
 import Data.Map.Util
-import Data.Monoid
+import Data.Monoid (Monoid(..))
 #if MIN_VERSION_base(4,9,0)
-import Data.Semigroup
+import Data.Semigroup (Semigroup(..))
+#endif
+#if !(MIN_VERSION_base(4,8,0))
+import Control.Applicative ((<$>))
+import Data.Traversable
 #endif
 import Prelude hiding (filter, lookup, null)
 import qualified Data.Map as M
@@ -67,7 +71,8 @@ instance (Data k, Data a, Ord k) => Data (OMap k a) where
 		1 -> k (z fromList)
 		_ -> error "gunfold"
 	dataTypeOf _   = oMapDataType
-	dataCast2      = gcast2
+	-- dataCast2 /must/ be eta-expanded in order to build on GHC 7.8.
+	dataCast2 f    = gcast2 f
 
 fromListConstr :: Constr
 fromListConstr = mkConstr oMapDataType "fromList" [] Prefix

--- a/Data/Set/Ordered.hs
+++ b/Data/Set/Ordered.hs
@@ -36,9 +36,9 @@ import Data.Foldable (Foldable, foldl', foldMap, foldr, toList)
 import Data.Function (on)
 import Data.Map (Map)
 import Data.Map.Util
-import Data.Monoid
+import Data.Monoid (Monoid(..))
 #if MIN_VERSION_base(4,9,0)
-import Data.Semigroup
+import Data.Semigroup (Semigroup(..))
 #endif
 import Data.Set (Set) -- so the haddocks link to the right place
 import Prelude hiding (filter, foldr, lookup, null)
@@ -63,7 +63,8 @@ instance (Data a, Ord a) => Data (OSet a) where
 		1 -> k (z fromList)
 		_ -> error "gunfold"
 	dataTypeOf _   = oSetDataType
-	dataCast1      = gcast1
+	-- dataCast1 /must/ be eta-expanded in order to build on GHC 7.8.
+	dataCast1 f    = gcast1 f
 
 fromListConstr :: Constr
 fromListConstr = mkConstr oSetDataType "fromList" [] Prefix

